app/src/main/java/ch/sheremet/dasfoo/rover/android/client/MainActivity.java

```diff
@@ -32,116 +35,142 @@
+                    || !host.equals(mGrpcConnection.getHost())
+                    || port != mGrpcConnection.getPort()) {
+                mGrpcConnection = new GrpcConnection(host, port);
+                // Android relies on a security Provider to provide secure network communications.
+                // It verifies that the security provider is up-to-date.
+                ProviderInstaller.installIfNeededAsync(MainActivity.this, this);
+            }
         }
 
         @Override
         protected void onPreExecute() {
-            mChannel = ManagedChannelBuilder.forAddress(mHost, mPort)
-                    .usePlaintext(true)
-                    .build();
-            mStub = RoverServiceGrpc.newBlockingStub(mChannel);
+            super.onPreExecute();
+            enableButtons(Boolean.FALSE);
         }
 
         @Override
         protected String doInBackground(final AbstractGrpcTaskExecutor... params) {
-            return params[0].execute(mStub);
+            return params[0].execute(mGrpcConnection.getStub());
+        }
+
+        @Override
+        protected void onPostExecute(final String result) {
+            mResultText.setText(result);
+            enableButtons(Boolean.TRUE);
+        }
+
+        @Override
+        public void onProviderInstalled() {
+            Toast toast = Toast.makeText(MainActivity.this, PROVIDER_INSTALLED, Toast.LENGTH_SHORT);
+            toast.show();
         }
 
         @Override
-        protected void onPostExecute(String result) {
-            try {
-                if (result == null) mResultText.setText(R.string.getting_null_result_text);
-                mResultText.setText(result);
-                mChannel.shutdown().awaitTermination(1, TimeUnit.SECONDS);
-                mListener.onGrpcTaskCompleted();
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
+        public void onProviderInstallFailed(final int errorCode, final Intent intent) {
+            GoogleApiAvailability googleAPI = GoogleApiAvailability.getInstance();
+            if (googleAPI.isUserResolvableError(errorCode)) {
+                googleAPI.getErrorDialog(MainActivity.this, errorCode,
+                        PLAY_SERVICES_RESOLUTION_REQUEST).show();
+                return;
             }
+            AlertDialog.Builder dialog = new AlertDialog.Builder(MainActivity.this);
+            dialog.setMessage(PROVIDER_NOT_INSTALLED);
+            dialog.setCancelable(false);
+            dialog.setPositiveButton(R.string.ok_button, new DialogInterface.OnClickListener() {
+                @Override
+                public void onClick(final DialogInterface dialog, final int which) {
+                    dialog.dismiss();
+                    Intent intent = new Intent(Intent.ACTION_MAIN);
+                    intent.addCategory(Intent.CATEGORY_HOME);
+                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                    startActivity(intent);
```

Comment:
So this throws the user at the main screen. What happens if the user re-opens the application from the "recent" (square button) menu?